### PR TITLE
Tempus: Remove ParameterList from TimeStepControlStrategies

### LINKED ISSUE #8372: Tempus:  Remove ParameterList from TimeStepControlStrategy Internals
## Enhancement
@trilinos/tempus 

Remove all the internal uses of ParameterList from TimeStepControlStrategies. This means moving the variables in the TimeStepControlStrategy ParameterList to member data. TimeStepControlStrategies will not longer inherit from Teuchos::ParameterListAcceptor. However TimeStepcontrolStrategies can still be built from a ParameterList, and will still provide a valid ParameterList.

This is in support of improving the data flow transparency for the Tempus PLATR deliverables.

### PR DESCRIPTION
**Pull request by ccober6**
Remove ParameterList from internals of TimeStepControlStrategies:
 - TimeStepControlStrategyBasicVS
 - TimeStepControlStrategyComposite
 - TimeStepControlStrategyConstant
 - TimeStepControlStrategyIntegralController

 * Move ParameterList constructor to a non-member function,
   createTimeStepControlStrategy*().
 * Created a default constructor and a full member data
   constructor.
 * Made initialize() check member data settings to ensure
   they are valid, and set isInitialized_.
 * Added setNextTimeStep() as the replacement for getNextTimeStep(),
   which better indicates what the function does.  getNextTimeStep()
   is still avaiable but just calls setNextTimeStep().
 * Made Strategies more consistent.
 * Added a step type to strategies to ensure it match the TimeStepControl.
 * Added initialization checks to the strategies.
 * Ensure each strategy updates the time step AND the time.
   - TimeStepControlStrategyConstant now sets the time from
     the dt and time step index to avoid numerical round-off.
 * Added to the documentation to TimeStepControlStrategyIntegralController.
 * Added unit tests for TimeStepControl and Strategies.

@trilinos/tempus 

## Motivation
Removing ParamerLists from the internals of the TimeStepControl objects simplifies the code and clarifies the data transparency. 

* Closes #8372 

## Testing
Added multiple unit tests covering TimeStepControl and TimeStepControlStrategies.

### CODE REVIEW

File: packages/tempus/src/Tempus_TimeStepControlStrategy.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_TimeStepControlStrategy.hpp:49]: Compilation error here, uncomment the `integratorStatus`

Comment (ccober6) [reply packages/tempus/src/Tempus_TimeStepControlStrategy.hpp:49]: Yeah, I was compiling with Tempus_HIDE_DEPRECATED_CODE=ON, so it missed this.  I have fixed it.  The bigger problem is with EMPIRE and the embedded strategy in EMPIRE-Fluid.  My changes have serious broke it.  :(  So I need to rework the TimeStepControlStrategyIntergalController, and probably put back in the ParameterList along with these changes so I can switch EIMPIRE-Fluid over to the new implementation, and then remove the ParameterList in another PR.

File: packages/tempus/src/Tempus_TimeStepControl_decl.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_TimeStepControl_decl.hpp:167]: `setNumTimeSteps` only one not setting `isInitialized_` to false.

Comment (ccober6) [reply packages/tempus/src/Tempus_TimeStepControl_decl.hpp:167]: I think it is being set in the *impl.hpp:458.

**Comment by ccober6** [comment]
@sconde What would you think about when setTimeStepControlStartegy is called, it would set the TimeStepControl:stepType_ to what the strategy's stepType_ is?

**Comment by sconde** [comment]
Not in this PR, but with the changes, I'm having issue with `SolutionHistory<Scalar>::addWorkingState`.

In the `updateTime` block, `workingState` time is set to `time += dt`. When using App's TSCS like CFL and TSC is created using PL, the `initialTimeStep = 1` so `workingState-getTime() = 0 + 1` before the inegrator even calls `takeStep`. 

What took me an embarrassingly long time to debug is even if the TSCS updates `dt` and the stepper correctly take step size of `dt` say 0.004, the integrator thinks `time = 1` and not `time = 0 + 0.004` and the it essently "completes" the integration with only 1 time step if the final time is shorter than 1.

What I did was just set the `workingState` time to `currentState` time and not add the timeStep. Running more tests...

**Comment by sconde** [comment]
> @sconde What would you think about when setTimeStepControlStartegy is called, it would set the TimeStepControl:stepType_ to what the strategy's stepType_ is?

This is basically what I had to do in Drekar to make sure things don't fail. If we can make sure Tempus is doing this, I think it will avoid a lot of headaches in the future

File: packages/tempus/src/Tempus_TimeStepControl_impl.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_TimeStepControl_impl.hpp:725]: So this here is in the Variable stepping block. I'm not sure I understand the reasoning.

Comment (ccober6) [reply packages/tempus/src/Tempus_TimeStepControl_impl.hpp:725]: Yeah, this coding has changed with my current changes which have not been committed yet.   This is some of the coding that was simplified by going with a TimeStepControl only having a single Strategy.  So I would suggest waiting on my up coming commit.  

How urgent do you need these changes?  I have been talking Eric Cyr about when I can merge this PR, and we are waiting for Jason Gates to return from vacation.  So hopefully next week.

Comment (sconde) [reply packages/tempus/src/Tempus_TimeStepControl_impl.hpp:725]: Not urging at all

**Comment by ccober6** [comment]
@sconde Here is the latest version.  I think I am done except for handling any issues and/or mods from you or our apps during initial testing.  Please let me know of any changes you think might be good.

**Comment by ccober6** [comment]
@ikalash @michael-a-hansen @sconde 

Hi all,

Here is a PR that has made some big changes to the TimeStepControl and TimeStepControlStrategy's (the commit message has details on the changes, but maybe the biggest thing you will notice is that "Integrator Step Type" has been removed).  

Can you check if there are problems with your apps and this PR?  I can help (i.e., pair-program with you) to incorporate these changes, if you like.  Sidafa mentioned working through the SPARC mods, so Mike you may want to coordinate with him.

Let me know if you have any problems or questions.

Thanks, Curt

**Comment by ikalash** [comment]
Thanks for the heads up, Curt.  I will test this tomorrow and report back what I find.

**Comment by ikalash** [comment]
Just tested Albany.  We're getting a lot of failures with the following error: 

```
 Error, the parameter {name="Integrator Step Type",type="string",value="Constant"}
1:  in the parameter (sub)list "Albany Parameters->Piro->Tempus->Tempus Integrator->Time Step Control"
1:  was not found in the list of valid parameters!
```
It looks like ``Integrator Step Type`` has been removed from the Tempus list of valid parameters.  It should be an easy fix to fix the failures - just change the input file.  Let me know if there's something I need to be aware of regarding this.

**Comment by ikalash** [comment]
Actually, question: what is the syntax to specify variable time-stepping now?

**Comment by ccober6** [comment]
Morning @ikalash!  

So yes "Integrator Step Type" has been removed and "Step Type" is determined by the "Time Step Control Strategy" that is specified (i.e., "Constant" or "Variable").  The default "Time Step Control Strategy" is the "Constant" strategy (i.e., the strategy used if not explicit set).  An explicit use of the "Constant" "Time Step Control Strategy" is shown in Trilinos/packages/tempus/test/ForwardEuler/Tempus_ForwardEuler_SinCos.xml.

For variable time stepping, you can use "Basic VS" (see Trilinos/packages/tempus/test/ForwardEuler/Tempus_ForwardEuler_VanDerPol.xml) or "Integral Controller" (see Trilinos/packages/tempus/test/DIRK/Tempus_DIRK_VanDerPol.xml - parameters in this example use the default values because they are not specified in the xml).

Hopefully these examples are helpful.  Let me know if this is not clear, or if you have questions.

File: packages/tempus/src/Tempus_TimeStepControlStrategyComposite.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_TimeStepControlStrategyComposite.hpp:220]: I'm getting a compilation error with a missing `#include "Tempus_TimeStepControlStrategyConstant.hpp"` in this file

Comment (ccober6) [reply packages/tempus/src/Tempus_TimeStepControlStrategyComposite.hpp:220]: Hmm ... checking it out.

Comment (ccober6) [reply packages/tempus/src/Tempus_TimeStepControlStrategyComposite.hpp:220]: @sconde So my mac and cee builds did not see this.  Also the autotester did not pick up on this either.  What are you building on?  (FYI, I do not think it is any problem adding this include.  Just interesting that only your build had an issue.)

Comment (sconde) [reply packages/tempus/src/Tempus_TimeStepControlStrategyComposite.hpp:220]: Building on my workstation. And I'm building from a clean build directory.

Comment (ccober6) [reply packages/tempus/src/Tempus_TimeStepControlStrategyComposite.hpp:220]: Yeah, my builds to were from clean builds.  Hard to fix when you cannot reproduce the failures.  :(

**Comment by ikalash** [comment]
Ok, I have gotten Albany to work with this PR.  I have created branches for both Albany and Albany-LCM, which I will merge in once this PR goes in.  Note that I had to make a change to Piro to get things to work correctly, which I just merged into this branch of Trilinos.  You have the thumbs up from me to merge this PR in.  I'll merge the Albany/Albany-LCM ones in once this is merged in.

**Comment by ccober6** [comment]
Thanks @ikalash!  It is odd that I did not get any errors in piro when I tested after my changes?!  I turned on 
 ```
 -D Trilinos_ENABLE_Piro:BOOL=ON \
  -D Piro_ENABLE_TESTS:BOOL=ON \
```
Are there additionally examples?

**Comment by ikalash** [comment]
@ccober6 : that particular code was not tested in Piro itself.  Due to limited Albany funds, some of the Piro capabilities are tested on the Albany side not in Piro itself.  I realize that this is far from ideal...

**Comment by ccober6** [comment]
@ikalash Ahh... that explains it.  I was worried that I had missed some testing somehow.  To complicate things, right now I cannot reproduce the autotester failures, so I was worried I had something really broken!  :)

Thanks for testing this so quickly.  I am making some final changes for Sidafa and hopefully be able to get this merged shortly.

**Comment by ikalash** [comment]
@ccober6 : sounds great. I'll keep an eye out for this getting merged in so I can merge in the Albany PRs afterwards.

**Comment by ccober6** [comment]
@ikalash @michael-a-hansen @sconde  So I think we are close to merging.  Are things ready on the SPARC side?  

Also Sidafa can you review this again?  Thanks!

**Comment by ikalash** [comment]
Albany is good with this.

**Comment by ccober6** [comment]
@sconde Ok, I just made the changes we talked about yesterday.  We should be ready to merge now.  Can you review this again?  Thanks!

**Comment by sconde** [comment]
In StrategyComposite, could we note that the strategy is chosen by the parameter "Strategy Type" and no longer by "Name"?

**Review by sconde — APPROVED** [review]
Looks good to me now :)

Thank you for your effort on this

**Pull request merged**
